tools/ci/ci_agent.py

```diff
@@ -278,6 +312,68 @@ def _run_targets(targets, src, build_path, config):
     return result
 
 
+def _build_code_coverage(src, build_path, build_flags, c_flags, codecov_token):
+    result = {}
+    try:
+        target_result = result.setdefault("coverage", {})
+        target_build_result = target_result.setdefault("CodeCoverage", {})
+        out = _build_target("coverage", src, build_path, build_flags, c_flags)
+        log(out)
+        out = _run_cmd(f"gcovr -r . -x -o {build_path}/cobertura.xml")
+        log(out)
+
+        commit_id = os.environ.get("ghprbActualCommit") or ""
+
+        if codecov_token and commit_id:
+            log("\n----------------------------------------------------------------")
+            log("Upload Code Coverage report to codecov.io")
+            log("----------------------------------------------------------------")
+
+            params = {
+                "token": codecov_token,
+                "commit": commit_id,
+                "branch": "",
+                "build": "",
+                "build_url": "",
+                "name": "",
+                "tag": "",
+                "slug": "",
+                "service": "",
+                "flags": "",
+                "pr": "",
+                "job": "",
+                "cmd_args": "",
+            }
+            headers = {"content-type": "application/x-www-form-urlencoded"}
+            with open(f"{build_path}/cobertura.xml", "rb") as payload:
+                try:
+                    requests.post(
+                        "https://codecov.io/upload/v2",
+                        data=payload,
+                        verify=True,
+                        headers=headers,
+                        params=params,
+                    )
+                except requests.exceptions.RequestException as err:
+                    log(f"Error calling codecov.io: {err}")
+        else:
+            if not commit_id:
+                log(
+                    "Please provide commit id, if you want to upload coverage report to codecov.io"
+                )
+            if not codecov_token:
+                log(
+                    "Provide '--codecov-token' in commandline, \
+                    if you want to upload coverage report to codecov.io"
+                )
```

Comment:
These appear to also be in either the environment or the config file?

Reply:
yes it is present in config file